imv: 4.1.0 -> 4.2.0

sternenseemann opened the pull request:
* switch build system to meson
* enable all backends by default

Open questions:

* imv prints a warning from the tiff backend everytime a non tiff file
  is opened: Is this normal? Seems harmless enough though.
* ~~Should we make backends configurable / optional? I readded some
  backends which apparently were removed, but still given as an argument
  to the derivation.~~

Resolves #108185.

cc @markus1189 @rnhmjoj 

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`) (**only wayland**)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

sternenseemann committed:
imv: 4.1.0 -> 4.2.0

* switch build system to wayland
* enable all backends by adding the following to buildInputs as meson
  autodetects which backends are available.
  * libtiff
  * libheif
  * libpng

Open questions:

* imv prints a warning from the tiff backend everytime a non tiff file
  is opened: Is this normal? Seems harmless enough though.
* Should we make backends configurable / optional? I readded some
  backends which apparently were removed, but still given as an argument
  to the derivation.

Resolves #108185.

mdlayher reviewed (approved):
LGTM.

```
$ nixpkgs-review pr 108309
$ git -c fetch.prune=false fetch --force https://github.com/NixOS/nixpkgs master:refs/nixpkgs-review/0 pull/108309/head:refs/nixpkgs-review/1
remote: Enumerating objects: 10, done.
remote: Counting objects: 100% (10/10), done.
remote: Compressing objects: 100% (5/5), done.
remote: Total 12 (delta 5), reused 5 (delta 5), pack-reused 2
Unpacking objects: 100% (12/12), 13.63 KiB | 4.54 MiB/s, done.
From ssh://github.com/NixOS/nixpkgs
   4d0e996b71d..712830d101d  master                -> refs/nixpkgs-review/0
 + 9d0164d53d4...a1e9a94785c refs/pull/108309/head -> refs/nixpkgs-review/1  (forced update)
$ git worktree add /home/matt/.cache/nixpkgs-review/pr-108309/nixpkgs 712830d101d94ae29a2310d91b4d54c0132f5716
Preparing worktree (detached HEAD 712830d101d)
HEAD is now at 712830d101d Merge pull request #108310 from mdlayher/mdl-bump-corerad
$ git merge --no-commit a1e9a94785cf885487d823790aa82dc82fe5924c
Automatic merge went well; stopped before committing as requested
$ nix build --no-link --keep-going --option build-use-sandbox relaxed -f /home/matt/.cache/nixpkgs-review/pr-108309/build.nix
[3 built, 21 copied (127.9 MiB), 22.4 MiB DL]
https://github.com/NixOS/nixpkgs/pull/108309
1 package built:
imv

$ nix-shell /home/matt/.cache/nixpkgs-review/pr-108309/shell.nix

```

rnhmjoj commented:
You beat me to it: I have just been working on this: https://github.com/rnhmjoj/nixpkgs/commit/f288f7ac25e.

> imv prints a warning from the tiff backend everytime a non tiff file
is opened: Is this normal? Seems harmless enough though.

It doesn't in 4.1.0, it's probably a bug but imv seems to work just fine.

> Should we make backends configurable / optional? I readded some

I would, but It doesn't look like they can be configured at all: the readme says most are optional but meson seems to require them all: maybe it's a bug in the build system. In any case, I would make X11/wayland configurable: you can pick from my commit if you want.

sternenseemann commented:
> It doesn't in 4.1.0, it's probably a bug but imv seems to work just fine.

imv 4.1.0 is compiled without `libtiff` in nixpkgs.

rnhmjoj commented:
There's an upstream issue for it: https://github.com/eXeC64/imv/issues/252

SuperSandro2000 commented:
Result of `nixpkgs-review pr 108309` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>imv</li>
  </ul>
</details>

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review which does not build all packages (e.g. lumo, tensorflow or pytorch)
If you find some bugs or got suggestions for further things to search or run please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 108309` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>imv</li>
</ul>
</details>

sternenseemann committed:
imv: use substituteInPlace

sternenseemann committed:
imv: make every backend and window system configurable

The following new derivation inputs are added:
* withBackends: a list of all backends to enable. The valid names are
  the same as specified in imv's meson_options.txt. Default is to enable
  all 7 backends: freeimage, libtiff, libjpeg(_turbo), libpng, librsvg,
  libnsgif and libheif
* withWindowSystem: either all, x11 or wayland to describe window system
  support.

sternenseemann commented:
@rnhmjoj made imv (very) configurable via overrides in e635fca398bafbfe4f4f7cf4771d474c8e5b29b9

rnhmjoj commented on pkgs/applications/graphics/imv/default.nix:
```diff
@@ -1,61 +1,116 @@
+      # alias wayland too avoid infinite recursion
+      wl = wayland;
+    in rec {
+      x11 = {
```
Cool idea!

rnhmjoj commented on pkgs/applications/graphics/imv/default.nix:
```diff
@@ -1,61 +1,116 @@
 , libGLU
-, libheif
+, wayland
+, withFreeimage ? true
```
Uhm, these are a lot of flags. It would be easier if you could do something like
```
imv.override { formats = [ "jpeg" "png" "svg" ] };
```
It could be implemented with 
```
-Dlibjpeg=${boolToEnable (elem "jpeg" formats)}
```

sternenseemann replied:
Might make for a bit of ugly asserts, but I think we can rework this into an `backends` parameter.

sternenseemann replied:
Done in 5d57a6b4dde41022650f0a9a7ff6f87799090150. Error messages for wrong `withWindowSystem` or `withBackends` values should be a lot more helpful now thanks to `lib.assertOneOf`.

rnhmjoj commented on pkgs/applications/graphics/imv/default.nix:
```diff
@@ -1,61 +1,116 @@
+, withFreeimage ? true
+, freeimage
+, withLibtiff ? true
+, libtiff
```
I'm not sure if this practice has changed, but the pattern used to be
```
, withSomething ? true
, libsomething ? null
...
}:

assert withSomething -> libsomething != null;
```
so that the package can be called without `libsomething`.

sternenseemann replied:
That's not necessary since `nix` is evaluated lazily. The resulting derivation won't reference `libsomething` and if you absolutely want to you can do the following without an evaluation error:

```nix
imv.override {
  withFreeimage = false;
  freeimage = null;
}
```

Shouldn't be necessary though. This approach saves a lot of asserts in our case and makes the code simpler and easier to read and I don't know of any disadvantages.

SuperSandro2000 commented:
Why once of the sudden do we need to disable backends? Can we just build them all and be done?

rnhmjoj commented:
> Why once of the sudden do we need to disable backends? Can we just build them all and be done? The package is really overengineered right now.

Closure size, for instance: disabling GIF and SVG, which are rarely used, saves about 20MB, TIFF saves 40MB, and more for others. 

imv is a very minimal image viewer and was designed to have optional backends, so it's should definitely be a feature of the package.

rnhmjoj commented on pkgs/applications/graphics/imv/default.nix:
```diff
@@ -1,61 +1,110 @@
+assert lib.assertOneOf "withWindowSystem" withWindowSystem
+  (builtins.attrNames windowSystems);
+# check that every given backend is valid
+assert builtins.foldl' lib.and true (builtins.map
```
You can simplify this using the `all` function:
```
builtins.all 
  (b: lib.assertOneOf "each backed" b withBackends)
  (builtins.attrNames backends)
```

sternenseemann replied:
Thanks!

rnhmjoj commented on pkgs/applications/graphics/imv/default.nix:
```diff
@@ -1,61 +1,110 @@
 
+let
+  windowSystems = {
+    x11 = {
```
Maybe use `x11.buildInputs` to save one indent level.

sternenseemann replied:
We can actually remove the attrset altogether if we need to handle extra info later we can always change it since it's not a public api or anything.

rnhmjoj reviewed (approved):
Looks all good. Thank you!
